Logcollector unit tests failure

|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.4 - 4.5 | CMocka unit tests | Manager -Agent | - | - |

# Description

Currently Logcollector state UT are failing due missing errno value after fopen wrapper.

https://github.com/wazuh/wazuh/blob/bc45e63702ad696c848cb448fea794de66d96373/src/unit_tests/logcollector/test_state.c#L874-L891

<details>
<summary> Unit test execution </summary>

```
Running tests...
Test project /tmp/build/wazuh/src/unit_tests/build
      Start  1: test_start_agent
 1/60 Test  #1: test_start_agent ....................   Passed    0.03 sec
      Start  2: test_notify
 2/60 Test  #2: test_notify .........................   Passed    0.03 sec
      Start  3: test_agentd_state
 3/60 Test  #3: test_agentd_state ...................   Passed    0.03 sec
      Start  4: test_buffer
 4/60 Test  #4: test_buffer .........................   Passed    0.03 sec
      Start  5: test_logcollector
 5/60 Test  #5: test_logcollector ...................   Passed    0.04 sec
      Start  6: test_read_multiline_regex
 6/60 Test  #6: test_read_multiline_regex ...........   Passed    0.04 sec
      Start  7: test_localfile-config
 7/60 Test  #7: test_localfile-config ...............   Passed    0.03 sec
      Start  8: test_state
 8/60 Test  #8: test_state ..........................***Failed    0.15 sec
[==========] Running 36 test(s).
[ RUN      ] test_w_logcollector_state_init_fail_hash_create_global
Expected assertion merror_exit called occurred
[       OK ] test_w_logcollector_state_init_fail_hash_create_global
[ RUN      ] test_w_logcollector_state_init_fail_hash_create_interval
Expected assertion merror_exit called occurred
[       OK ] test_w_logcollector_state_init_fail_hash_create_interval
[ RUN      ] test_w_logcollector_state_init_fail_hash_setsize_global
Expected assertion merror_exit called occurred
[       OK ] test_w_logcollector_state_init_fail_hash_setsize_global
[ RUN      ] test_w_logcollector_state_init_fail_hash_setsize_interval
Expected assertion merror_exit called occurred
[       OK ] test_w_logcollector_state_init_fail_hash_setsize_interval
[ RUN      ] test_w_logcollector_state_init_ok
[       OK ] test_w_logcollector_state_init_ok
[ RUN      ] test_w_logcollector_state_get_null
[       OK ] test_w_logcollector_state_get_null
[ RUN      ] test_w_logcollector_state_get_non_null
[       OK ] test_w_logcollector_state_get_non_null
[ RUN      ] test__w_logcollector_generate_state_fail_get_node
[       OK ] test__w_logcollector_generate_state_fail_get_node
[ RUN      ] test__w_logcollector_generate_state_one_target
[       OK ] test__w_logcollector_generate_state_one_target
[ RUN      ] test__w_logcollector_generate_state_one_target_restart
[       OK ] test__w_logcollector_generate_state_one_target_restart
[ RUN      ] test__w_logcollector_state_update_file_new_data
[       OK ] test__w_logcollector_state_update_file_new_data
[ RUN      ] test__w_logcollector_state_update_file_update
[       OK ] test__w_logcollector_state_update_file_update
[ RUN      ] test__w_logcollector_state_update_file_fail_update
[       OK ] test__w_logcollector_state_update_file_fail_update
[ RUN      ] test_w_logcollector_state_update_file_null
[       OK ] test_w_logcollector_state_update_file_null
[ RUN      ] test_w_logcollector_state_update_file_ok
[       OK ] test_w_logcollector_state_update_file_ok
[ RUN      ] test__w_logcollector_state_update_target_get_file_stats_fail
[       OK ] test__w_logcollector_state_update_target_get_file_stats_fail
[ RUN      ] test__w_logcollector_state_update_target_find_target_fail
[       OK ] test__w_logcollector_state_update_target_find_target_fail
[ RUN      ] test__w_logcollector_state_update_target_find_target_ok
[       OK ] test__w_logcollector_state_update_target_find_target_ok
[ RUN      ] test__w_logcollector_state_update_target_dropped_true
[       OK ] test__w_logcollector_state_update_target_dropped_true
[ RUN      ] test__w_logcollector_state_update_target_OSHash_Update_fail
[       OK ] test__w_logcollector_state_update_target_OSHash_Update_fail
[ RUN      ] test__w_logcollector_state_update_target_OSHash_Add_fail
[       OK ] test__w_logcollector_state_update_target_OSHash_Add_fail
[ RUN      ] test_w_logcollector_state_update_target_null_path
[       OK ] test_w_logcollector_state_update_target_null_path
[ RUN      ] test_w_logcollector_state_update_target_null_target
[       OK ] test_w_logcollector_state_update_target_null_target
[ RUN      ] test_w_logcollector_state_update_target_ok
[       OK ] test_w_logcollector_state_update_target_ok
[ RUN      ] test_w_logcollector_generate_state_ok
[       OK ] test_w_logcollector_generate_state_ok
[ RUN      ] test_w_logcollector_state_dump_fail_open
[  ERROR   ] --- "(1103): Could not open file 'var/run/wazuh-logcollector.state' due to [(2)-(No such file or directory)]." != "(1103): Could not open file 'var/run/wazuh-logcollector.state' due to [(0)-(Success)]."
unit_tests/wrappers/wazuh/shared/debug_op_wrappers.c:61: error: Check of parameter formatted_msg, function __wrap__merror failed
/tmp/build/wazuh/src/unit_tests/logcollector/test_state.c:886: note: Expected parameter declared here
[   LINE   ] --- unit_tests/wrappers/wazuh/shared/debug_op_wrappers.c:61: error: Failure!
[  FAILED  ] test_w_logcollector_state_dump_fail_open
[ RUN      ] test_w_logcollector_state_dump_fail_write
[  ERROR   ] --- "(1110): Could not write file 'var/run/wazuh-logcollector.state' due to [(2)-(No such file or directory)]." != "(1110): Could not write file 'var/run/wazuh-logcollector.state' due to [(0)-(Success)]."
unit_tests/wrappers/wazuh/shared/debug_op_wrappers.c:61: error: Check of parameter formatted_msg, function __wrap__merror failed
/tmp/build/wazuh/src/unit_tests/logcollector/test_state.c:906: note: Expected parameter declared here
[   LINE   ] --- unit_tests/wrappers/wazuh/shared/debug_op_wrappers.c:61: error: Failure!
[  FAILED  ] test_w_logcollector_state_dump_fail_write
[ RUN      ] test_w_logcollector_state_dump_ok
[       OK ] test_w_logcollector_state_dump_ok
[ RUN      ] test_w_logcollector_state_main_bad_interval
[       OK ] test_w_logcollector_state_main_bad_interval
[ RUN      ] test_w_logcollector_state_main_ok
[       OK ] test_w_logcollector_state_main_ok
[ RUN      ] test__w_logcollector_state_delete_file_no_data
[       OK ] test__w_logcollector_state_delete_file_no_data
[ RUN      ] test__w_logcollector_state_delete_file_ok
[       OK ] test__w_logcollector_state_delete_file_ok
[ RUN      ] test_w_logcollector_state_delete_file_fpath_NULL
[       OK ] test_w_logcollector_state_delete_file_fpath_NULL
[ RUN      ] test_w_logcollector_state_delete_file_global
[       OK ] test_w_logcollector_state_delete_file_global
[ RUN      ] test_w_logcollector_state_delete_file_interval
[       OK ] test_w_logcollector_state_delete_file_interval
[ RUN      ] test_w_logcollector_state_delete_file_global_interval
[       OK ] test_w_logcollector_state_delete_file_global_interval
[==========] 36 test(s) run.
[  PASSED  ] 34 test(s).
[  FAILED  ] 2 test(s), listed below:
[  FAILED  ] test_w_logcollector_state_dump_fail_open
[  FAILED  ] test_w_logcollector_state_dump_fail_write

 2 FAILED TEST(S)

=================================================================
==5027==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 10 byte(s) in 1 object(s) allocated from:
    #0 0x7fb43d2e2c18 in __interceptor_realloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:164
    #1 0x563446ce63f3 in w_logcollector_state_dump logcollector/state.c:113
    #2 0x563446cd3218 in test_w_logcollector_state_dump_fail_write /tmp/build/wazuh/src/unit_tests/logcollector/test_state.c:913
    #3 0x7fb43c9274e0  (/lib/x86_64-linux-gnu/libcmocka.so.0+0x54e0)

Direct leak of 10 byte(s) in 1 object(s) allocated from:
    #0 0x7fb43d2e2c18 in __interceptor_realloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:164
    #1 0x563446ce63f3 in w_logcollector_state_dump logcollector/state.c:113
    #2 0x563446cd2ee9 in test_w_logcollector_state_dump_fail_open /tmp/build/wazuh/src/unit_tests/logcollector/test_state.c:890
    #3 0x7fb43c9274e0  (/lib/x86_64-linux-gnu/libcmocka.so.0+0x54e0)

SUMMARY: AddressSanitizer: 20 byte(s) leaked in 2 allocation(s).

      Start  9: test_lccom
 9/60 Test  #9: test_lccom ..........................   Passed    0.04 sec
      Start 10: test_macos_log
10/60 Test #10: test_macos_log ......................   Passed    0.03 sec
      Start 11: test_read_macos
11/60 Test #11: test_read_macos .....................   Passed    0.04 sec
      Start 12: test_execd
12/60 Test #12: test_execd ..........................   Passed    0.03 sec
      Start 13: test_wm_github
13/60 Test #13: test_wm_github ......................   Passed    0.05 sec
      Start 14: test_wm_office365
14/60 Test #14: test_wm_office365 ...................   Passed    0.08 sec
      Start 15: test_create_db
15/60 Test #15: test_create_db ......................   Passed    0.06 sec
      Start 16: test_syscom
16/60 Test #16: test_syscom .........................   Passed    0.02 sec
      Start 17: test_fim_diff_changes
17/60 Test #17: test_fim_diff_changes ...............   Passed    0.04 sec
      Start 18: test_run_realtime
18/60 Test #18: test_run_realtime ...................   Passed    0.05 sec
      Start 19: test_syscheck_config
19/60 Test #19: test_syscheck_config ................   Passed    0.06 sec
      Start 20: test_syscheck
20/60 Test #20: test_syscheck .......................   Passed    0.02 sec
      Start 21: test_fim_sync
21/60 Test #21: test_fim_sync .......................   Passed    0.09 sec
      Start 22: test_run_check
22/60 Test #22: test_run_check ......................   Passed    0.10 sec
      Start 23: test_fim_db
23/60 Test #23: test_fim_db .........................   Passed    0.05 sec
      Start 24: test_fim_db_files
24/60 Test #24: test_fim_db_files ...................   Passed    0.05 sec
      Start 25: test_audit_healthcheck
25/60 Test #25: test_audit_healthcheck ..............   Passed    0.04 sec
      Start 26: test_audit_rule_handling
26/60 Test #26: test_audit_rule_handling ............   Passed    0.04 sec
      Start 27: test_syscheck_audit
27/60 Test #27: test_syscheck_audit .................   Passed    0.05 sec
      Start 28: test_audit_parse
28/60 Test #28: test_audit_parse ....................   Passed    0.05 sec
      Start 29: test_list_op
29/60 Test #29: test_list_op ........................   Passed    0.03 sec
      Start 30: test_file_op
30/60 Test #30: test_file_op ........................   Passed    0.02 sec
      Start 31: test_integrity_op
31/60 Test #31: test_integrity_op ...................   Passed    0.02 sec
      Start 32: test_rbtree_op
32/60 Test #32: test_rbtree_op ......................   Passed    0.02 sec
      Start 33: test_validate_op
33/60 Test #33: test_validate_op ....................   Passed    0.03 sec
      Start 34: test_string_op
34/60 Test #34: test_string_op ......................   Passed    0.02 sec
      Start 35: test_expression
35/60 Test #35: test_expression .....................   Passed    0.02 sec
      Start 36: test_version_op
36/60 Test #36: test_version_op .....................   Passed    0.03 sec
      Start 37: test_queue_op
37/60 Test #37: test_queue_op .......................   Passed    0.02 sec
      Start 38: test_queue_linked_op
38/60 Test #38: test_queue_linked_op ................   Passed    0.02 sec
      Start 39: test_agent_op
39/60 Test #39: test_agent_op .......................   Passed    0.02 sec
      Start 40: test_enrollment_op
40/60 Test #40: test_enrollment_op ..................   Passed    0.03 sec
      Start 41: test_time_op
41/60 Test #41: test_time_op ........................   Passed    0.02 sec
      Start 42: test_buffer_op
42/60 Test #42: test_buffer_op ......................   Passed    0.02 sec
      Start 43: test_utf8_op
43/60 Test #43: test_utf8_op ........................   Passed    0.02 sec
      Start 44: test_log_builder
44/60 Test #44: test_log_builder ....................   Passed    0.02 sec
      Start 45: test_custom_output_search_replace
45/60 Test #45: test_custom_output_search_replace ...   Passed    0.02 sec
      Start 46: test_syscheck_op
46/60 Test #46: test_syscheck_op ....................   Passed    0.03 sec
      Start 47: test_audit_op
47/60 Test #47: test_audit_op .......................   Passed    0.02 sec
      Start 48: test_privsep_op
48/60 Test #48: test_privsep_op .....................   Passed    0.02 sec
      Start 49: test_mq_op
49/60 Test #49: test_mq_op ..........................   Passed    0.02 sec
      Start 50: test_remoted_op
50/60 Test #50: test_remoted_op .....................   Passed    0.04 sec
      Start 51: test_atomic
51/60 Test #51: test_atomic .........................   Passed    0.02 sec
      Start 52: test_url
52/60 Test #52: test_url ............................   Passed    0.03 sec
      Start 53: test_sysinfo_utils
53/60 Test #53: test_sysinfo_utils ..................   Passed    0.02 sec
      Start 54: test_os_xml
54/60 Test #54: test_os_xml .........................   Passed    0.07 sec
      Start 55: test_os_regex
55/60 Test #55: test_os_regex .......................   Passed    0.02 sec
      Start 56: test_os_regex_execute
56/60 Test #56: test_os_regex_execute ...............   Passed    0.04 sec
      Start 57: test_os_zlib
57/60 Test #57: test_os_zlib ........................   Passed    0.02 sec
      Start 58: test_os_net
58/60 Test #58: test_os_net .........................   Passed    0.02 sec
      Start 59: test_fluentd_forwarder
59/60 Test #59: test_fluentd_forwarder ..............   Passed    0.04 sec
      Start 60: test_active-response
60/60 Test #60: test_active-response ................   Passed    0.02 sec

98% tests passed, 1 tests failed out of 60

Total Test time (real) =   2.26 sec

The following tests FAILED:
	  8 - test_state (Failed)
Errors while running CTest
make: *** [Makefile:71: test] Error 8
```

</details>

## RCA

Failure seems to be related to using an unset `errno` according to wrapped `fopen` mock expected behavior.

--- a/src/unit_tests/logcollector/test_state.c
+++ b/src/unit_tests/logcollector/test_state.c
@@ -883,9 +883,10 @@ void test_w_logcollector_state_dump_fail_open(void ** state) {
     expect_string(__wrap_fopen, mode, "w");
     will_return(__wrap_fopen, NULL);
 
-    expect_string(__wrap__merror, formatted_msg,
-                  "(1103): Could not open file "
-                  "'var/run/wazuh-logcollector.state' due to [(0)-(Success)].");
+    const char * error_msg = "(1103): Could not open file "
+                             "'" LOGCOLLECTOR_STATE "' due to";
+
+    expect_memory(__wrap__merror, formatted_msg, error_msg, strlen(error_msg));
 
     w_logcollector_state_dump();
 }
@@ -903,9 +904,10 @@ void test_w_logcollector_state_dump_fail_write(void ** state) {
     will_return(__wrap_fopen, (FILE *) 100);
     will_return(__wrap_fwrite, 0);
 
-    expect_string(__wrap__merror, formatted_msg,
-                  "(1110): Could not write file "
-                  "'var/run/wazuh-logcollector.state' due to [(0)-(Success)].");
+    const char * error_msg = "(1110): Could not write file "
+                             "'" LOGCOLLECTOR_STATE "' due to";
+
+    expect_memory(__wrap__merror, formatted_msg, error_msg, strlen(error_msg));
 
     expect_value(__wrap_fclose, _File, (FILE *) 100);
     will_return(__wrap_fclose, 0);

--- a/src/unit_tests/wazuh_db/test_wdb_global_helpers.c
+++ b/src/unit_tests/wazuh_db/test_wdb_global_helpers.c
@@ -254,7 +254,9 @@ void test_wdb_create_agent_db_error_changing_owner(void **state)
     expect_value(__wrap_chown, __group, 0);
     will_return(__wrap_chown, OS_INVALID);
     will_return(__wrap_strerror, "error");
-    expect_string(__wrap__merror, formatted_msg, "(1135): Could not chown object 'var/db/agents/001-agent1.db' due to [(0)-(error)].");
+
+    const char * error_msg = "(1135): Could not chown object 'var/db/agents/001-agent1.db' due to";
+    expect_memory(__wrap__merror, formatted_msg, error_msg, strlen(error_msg));
 
     ret = wdb_create_agent_db(agent_id, agent_name);
 
@@ -299,7 +301,9 @@ void test_wdb_create_agent_db_error_changing_mode(void **state)
     expect_string(__wrap_chmod, path, "var/db/agents/001-agent1.db");
     will_return(__wrap_chmod, OS_INVALID);
     will_return(__wrap_strerror, "error");
-    expect_string(__wrap__merror, formatted_msg, "(1127): Could not chmod object 'var/db/agents/001-agent1.db' due to [(0)-(error)].");
+
+    const char * error_msg = "(1127): Could not chmod object 'var/db/agents/001-agent1.db' due to";
+    expect_memory(__wrap__merror, formatted_msg, error_msg, strlen(error_msg));
 
     ret = wdb_create_agent_db(agent_id, agent_name);
 
@@ -2177,12 +2181,16 @@ void test_wdb_remove_agent_db_error_removing_db_shm_wal(void **state) {
     expect_string(__wrap_remove, filename, "var/db/agents/001-agent1.db-shm");
     will_return(__wrap_remove, OS_INVALID);
     will_return(__wrap_strerror, "error");
-    expect_string(__wrap__mdebug2, formatted_msg, "(1129): Could not unlink file 'var/db/agents/001-agent1.db-shm' due to [(0)-(error)].");
+
+    const char * error_shm = "(1129): Could not unlink file 'var/db/agents/001-agent1.db-shm' due to";
+    expect_memory(__wrap__mdebug2, formatted_msg, error_shm, strlen(error_shm));
 
     expect_string(__wrap_remove, filename, "var/db/agents/001-agent1.db-wal");
     will_return(__wrap_remove, OS_INVALID);
     will_return(__wrap_strerror, "error");
-    expect_string(__wrap__mdebug2, formatted_msg, "(1129): Could not unlink file 'var/db/agents/001-agent1.db-wal' due to [(0)-(error)].");
+
+    const char * error_wal = "(1129): Could not unlink file 'var/db/agents/001-agent1.db-wal' due to";
+    expect_memory(__wrap__mdebug2, formatted_msg, error_wal, strlen(error_wal));
 
     ret = wdb_remove_agent_db(id, name);
 

